Fix multi-commodity transactions, add preferences

abachma2 opened the pull request:
This PR has two primary additions:
- Input commodity preferences are added (new feature)
- Transaction issues are fixed when multiple commodities are requested (a bug fix, for a bug discovered while adding the feature)

Test and test-related files have been updated to test the new functionality and ensure the bug is fixed. There are two tests that are failing because of a bug with resource IDs for the spent fuel. This bug (Issue #9) will be fixed in a later PR. The 2 tests regarding transactions are both passing, and those are the tests for the feature and bug fix in this PR.

abachma2 commented:
Yes. This is ready for review.

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
     def tock(self):
-        print("World \n")
+        '''
+        Logic to implement at the tock phase of each 
+        time step. 
+
+        If the prototype is retired, then nothing happens. 
+
+        If it's after a cycle ends and the refueling time has passed,
+        the core is full, and fuel has been discharged, then 
+        the discharged variable is changed to false and the 
+        cycle length counter is restarted.
+
+        If it's the beginning of a new cycle and the core is full, 
+        then a cycle start is recorded. 
+
+        If it's in the middle of a cycle and the core is full, the 
+        the power_cap value is recorded as power generated. If these 
+        conditions aren't met, then a power of 0 is recorded. 
+
+        If it's in the middle of a cycle or the core is full, then 
+        the cycle duration counter increases by one. 
```
I like the way you've set up these doc strings

abachma2 replied:
Thanks! It's mostly me trying to keep track of actions as I translated things from C++ to python

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+            self.cycle_step = 0
+        
+        if (self.cycle_step == 0) and (self.core.count == self.n_assem_core):
+            #self.record("CYCLE_START", "")
```
Remove comment

abachma2 replied:
See above about `self.record`

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        submit no bids for materials.
+        '''
+        port = []
+        n_assem_order = self.n_assem_core - self.core.count + self.n_assem_fresh + self.fresh_fuel.count
```
Based on the docstring, shouldn't this be `... - self.fresh_fuel.count`?

abachma2 replied:
Yes it should be. Thank you for catching that.

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+
+        Add a constraint of the total quantity of the commodity available.
+        Create a bid portfolio for each request that can be met. 
+        '''
```
Consider adding a description of the `requests` parameter to the docstring

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        print("time:", self.context.time, "respond", len(bids), "assemblies")
+        return port
+
+    def get_material_trades(self, trades): #phase 5.1
```
Phase 2 jumped to phase 5.1?

abachma2 replied:
Yes. The DRE has 5 different phases. Phase 3 is a preference adjustment phase, which I am not employing here, and Phase 4 is the matching of requests and bids and doesn't require a method in the archetype. More information can be found [here](https://fuelcycle.org/arche/dre).

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        composition of the trade.
+
+        Then trade the materials from the spent fuel inventory. 
+        '''
```
Consider adding information about the parameter `trades` to this docstring and the returned `responses`

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        print("time:", self.context.time, "accept", n_load, "assemblies")
+        if n_load > 0:
+            ss = str(n_load) + " assemblies"
+            #self.record("LOAD", ss)
```
Remove comment

abachma2 replied:
See above about `self.record`.

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        if n == 0:
+            return
+        ss = str(n) + " assemblies"
+        #self.record("LOAD", ss)
```
```suggestion

```

abachma2 replied:
See above about `self.record`

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        for ii in range(len(old)):
+            self.core.push(old[ii]) 
+        ss = str(len(old)) + " assemblies"
+        #self.record("TRANSMUTE", ss)
```
```suggestion

```

abachma2 replied:
See above about `self.record`

LukeSeifert commented on tests/integration_tests/test_depletereactor.py:
```diff
@@ -0,0 +1,172 @@
+        spec: str
+            value to find in a column
+        a: table
+            database table to searc
```
```suggestion
            database table to search
```

LukeSeifert reviewed (commented):
Overall looks good, I think there are some minor changes you can make which could polish things up.

abachma2 commented on examples/simple.xml:
```diff
@@ -39,14 +102,53 @@
               <number>1</number>
               <prototype>OneReactor</prototype>
             </entry>  
+            <!--entry>
+              <number>1</number>
+              <prototype>TwoReactor</prototype>
+            </entry-->  
```
I am benchmarking a lot of the functionality against the Cycamore Reactor archetype, so this was an easy way for me to switch between running the simulation with the OpenMCyclus and Cycamore archetypes. I will remove this section because I ended up creating another file that runs with the Cycamore archetype, so this is no longer needed.

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        default=0
+    )
+
+    refuel_time = ts.Int(
+        doc = "Time steps for refueling",
+        tooltip="Time steps for refueling",
+        uilabel="refueltime",
+        default = 0
+    )
+
+    n_assem_core = ts.Int(
+        doc = "Number of assemblies in a core",
+        tooltip = "Number of assemblies in a core",
+        uilabel = "n_assem_core",
+        default=0
+    )
```
Will do!

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        print("time:", self.context.time, "tick")
+        if self.retired():
+            print("time:", self.context.time, "retired")
+        #    self.record("RETIRED", "")
```
The comments for the `self.record` method are intended to be eventual functionality to record to the database when certain things happen, such as starting a new cycle or retiring a reactor.

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+            print("time:", self.context.time, self.fresh_fuel.count)
+            self.load()
+
+        #lib.Logger('5', str("DepleteReactor" + str(self.power_cap) + "is ticking"))
```
Similar to `self.record` the `lib.Logger` is intended to be functional at some point. This functionality would write to the command line when cyclus is called in verbose mode.

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+            return 0
+
+    def discharge(self):
+        '''
+        '''
```
Will add.

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        '''
+        npop = min(self.n_assem_batch, self.core.count)
+        if (self.n_assem_spent - self.spent_fuel.count) < npop:
+            #self.record("DISCHARGE", "failed")
```
See above about `self.record`

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        return
+
+    def record(self, event, val):
+        '''
+        Record a reactor event to the output database with the 
+        given name and note val
+
+        Parameters:
+        -----------
+        event: str 
+            name of event
+        val: str
+            value of event
+        '''
```
Will do. All of the parameters and outputs of these functions are never seen by the user and are passed between the archetype and the cyclus kernel, so that hasn't been a high priority.

abachma2 commented on openmcyclus/DepleteReactor.py:
```diff
@@ -8,8 +11,603 @@ class DepleteReactor(Facility):
+        '''
+        For each assembly in the spent fuel inventory, 
+        get the commodity name of it, and push the material 
+        back to the spent fuel inventory (??).
```
A lot of the documentation is notes to myself about what is happening in the Cycamore Reactor archetype. So these question marks mean that I don't fully understand something in that archetype.

abachma2 commented:
Thanks for the review @LukeSeifert. I'll work on making the documentation better. I have made some of the other changes requested.

abachma2 commented:
I have found a bug in the feature added through this PR. The requested material is based more on their ordering in the input file than the defined preferences. This results in errors in how much of the material is accepted if there is not enough to meet a full request. I am working on fixing this bug.

abachma2 commented:
I have updated the doc strings to be more comprehensive. I have also updated the bug for the preferences described in the previous comment. The bug was because of a lack of feature in the Python API in cyclus, so that has been [added in](https://github.com/cyclus/cyclus/pull/1584).

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -347,97 +420,140 @@ def get_material_bids(self, requests): # phase 2
+        print(trades)
         for ii in range(len(trades)):
             print("time:", self.context.time, "number of trades:", len(trades))
+            print(trades[ii])
```
Consider adding more detail (such as `print(f'Trade {ii}: {trades[ii]}')` or something similar) so the information is easier to read

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -347,97 +420,140 @@ def get_material_bids(self, requests): # phase 2
             responses[trades[ii]] = mat
             self.resource_indexes.pop(mat.obj_id)
         self.push_spent(mats)
+        print(responses)
```
Consider adding more detail here as well

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -347,97 +420,140 @@ def get_material_bids(self, requests): # phase 2
         '''
-        n_load = min(len(responses), self.n_assem_core - self.core.count)
+        print("time:", self.context.time, "responses:", responses)
+        # min(len(responses), self.n_assem_core - self.core.count)
```
Removable?

LukeSeifert commented on openmcyclus/DepleteReactor.py:
```diff
@@ -347,97 +420,140 @@ def get_material_bids(self, requests): # phase 2
-            #self.record("LOAD", ss)
+            # self.record("LOAD", ss)
         for trade in responses:
+            print(trade.request.commodity, trade.request.preference, trade.amt)
```
Consider adding more detail to this print statement

LukeSeifert reviewed (approved):
Very good changes! Overall I just saw some print statements that I thought could use more detail in the output, but it's up to you if you think adding that detail makes sense or not. Let me know if you would like me to go ahead and merge.

abachma2 commented:
Thanks for the review. Those print statements were meant for debugging for development and aren't needed in the final version. I'll remove them.

abachma2 commented:
The extra print statements have been removed and pep8 formatting has been enforced.
